In [9]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OrdinalEncoder
from keras.models import Sequential
from keras.layers import Dense


In [13]:
def load_dataset(filename):
    data = pd.read_csv(filename, header=None)
    dataset = data.values
    #split to input x and output y
    x = dataset[:,:-1]
    y = dataset[:,-1]

    x = x.astype(str) #format values as string
    y = y.reshape((len(y),1)) #reshape target to 2d array

    return x, y
    

In [16]:
x, y = load_dataset('breast-cancer.csv')
#split intro train and test sets
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.33, random_state=1)

In [ ]:
print('Train', x_train.shape, y_train.shape)
print('Test', x_test.shape, y_test.shape)

Train (191, 9) (191, 1)
Test (95, 9) (95, 1)


We have 191 samples for training and 95 for testing.

In [ ]:
def input_prep(x_train, x_test):
    oe = sklpre.OrdinalEncoder()
    oe.fit(x_train)
    x_train_enc = oe.transform(x_train)
    x_test_enc = oe.transform(x_test)
    return x_train_enc, x_test_enc

In [ ]:
def output_prep(y_train, y_test):
    le = sklpre.LabelEncoder()
    le.fit(y_train)
    y_train_enc = le.transform(y_train)
    y_test_enc = le.transform(y_test)
    return y_train_enc, y_test_enc

In [ ]:
x_train_enc, x_test_enc = input_prep(x_train, x_test)
y_train_enc, y_test_enc = output_prep(y_train, y_test)

Neural Network Model:

In [ ]:
# define the model
model = Sequential()
model.add(Dense(10, input_dim=x_train_enc.shape[1], activation='relu', kernel_initializer='he_normal'))
model.add(Dense(1, activation='sigmoid'))
# compile the keras model
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
# fit the keras model on the dataset
model.fit(x_train_enc, y_train_enc, epochs=100, batch_size=16, verbose=2)
# evaluate the keras model
_, accuracy = model.evaluate(x_test_enc, y_test_enc, verbose=0)
print('Accuracy: %.2f' % (accuracy*100))

Epoch 1/100
12/12 - 1s - loss: 0.8825 - accuracy: 0.3194 - 534ms/epoch - 45ms/step
Epoch 2/100
12/12 - 0s - loss: 0.8202 - accuracy: 0.3822 - 15ms/epoch - 1ms/step
Epoch 3/100
12/12 - 0s - loss: 0.7730 - accuracy: 0.4398 - 15ms/epoch - 1ms/step
Epoch 4/100
12/12 - 0s - loss: 0.7358 - accuracy: 0.5131 - 15ms/epoch - 1ms/step
Epoch 5/100
12/12 - 0s - loss: 0.7090 - accuracy: 0.6021 - 17ms/epoch - 1ms/step
Epoch 6/100
12/12 - 0s - loss: 0.6872 - accuracy: 0.6283 - 19ms/epoch - 2ms/step
Epoch 7/100
12/12 - 0s - loss: 0.6696 - accuracy: 0.6492 - 18ms/epoch - 2ms/step
Epoch 8/100
12/12 - 0s - loss: 0.6577 - accuracy: 0.6649 - 16ms/epoch - 1ms/step
Epoch 9/100
12/12 - 0s - loss: 0.6459 - accuracy: 0.6754 - 16ms/epoch - 1ms/step
Epoch 10/100
12/12 - 0s - loss: 0.6363 - accuracy: 0.6806 - 16ms/epoch - 1ms/step
Epoch 11/100
12/12 - 0s - loss: 0.6275 - accuracy: 0.6963 - 15ms/epoch - 1ms/step
Epoch 12/100
12/12 - 0s - loss: 0.6205 - accuracy: 0.6963 - 15ms/epoch - 1ms/step
Epoch 13/100
12/12 - 0s